radixdlt-core/radixdlt/src/integration/java/com/radixdlt/integration/api/actors/TreeOfSubstatesChecker.java

```diff
@@ -0,0 +1,130 @@
+ * of any economic or other opportunity of whatsoever nature or howsoever arising), arising
+ * out of or in connection with (without limitation of any use, misuse, of any ledger system
+ * or use made or its functionality or any performance or operation of any code or protocol
+ * caused by bugs or programming or logic errors or otherwise);
+ *
+ * A. any offer, purchase, holding, use, sale, exchange or transmission of any
+ * cryptographic keys, tokens or assets created, exchanged, stored or arising from any
+ * interaction with the Work;
+ *
+ * B. any failure in a transmission or loss of any token or assets keys or other digital
+ * artefacts due to errors in transmission;
+ *
+ * C. bugs, hacks, logic errors or faults in the Work or any communication;
+ *
+ * D. system software or apparatus including but not limited to losses caused by errors
+ * in holding or transmitting tokens by any third-party;
+ *
+ * E. breaches or failure of security including hacker attacks, loss or disclosure of
+ * password, loss of private key, unauthorised use or misuse of such passwords or keys;
+ *
+ * F. any losses including loss of anticipated savings or other benefits resulting from
+ * use of the Work or any changes to the Work (however implemented).
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+package com.radixdlt.integration.api.actors;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import com.radixdlt.environment.deterministic.MultiNodeDeterministicRunner;
+import com.radixdlt.integration.api.DeterministicActor;
+import com.radixdlt.tree.substate.BerkeleySubStateStore;
+import com.radixdlt.utils.Bytes;
+import com.radixdlt.utils.Longs;
+import com.sleepycat.je.Cursor;
+import com.sleepycat.je.Database;
+import com.sleepycat.je.DatabaseEntry;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Random;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public final class TreeOfSubstatesChecker implements DeterministicActor {
+  private static final Logger logger = LogManager.getLogger();
+
+  @Override
+  public String execute(MultiNodeDeterministicRunner runner, Random random) throws Exception {
+
+    Long highestEpoch = 0L;
+    Map<Long, byte[]> hashesAtEpochNumber = new HashMap<>();
+
+    for (int i = 0; i < runner.getSize(); i++) {
+      var injector = runner.getNode(i);
+
+      var berkeleySubStateStore = injector.getInstance(BerkeleySubStateStore.class);
+      Database epochRootHashDatabase = berkeleySubStateStore.getEpochRootHashDatabase();
```

Comment:
Would rather have the retrieval of the epoch rootHash implemented in `BerkeleySubStateStore`

Reply:
Isn't it more efficient in the way it is implemented? The code needs the cursor to get the previous epoch, potentially many times. The `BerkeleySubStateStore` would need to open and close the cursor many times.